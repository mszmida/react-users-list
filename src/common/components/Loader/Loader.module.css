.component {
  text-align: center;
  margin: 50px;
}

.loaderDualRing {
  display: inline-block;
  width: 80px;
  height: 80px;
}

.loaderDualRing:after {
  content: ' ';
  display: block;
  width: 64px;
  height: 64px;
  margin: 8px;
  border-radius: 50%;
  border: 6px solid black;
  border-color: black transparent black transparent;
  animation: dual-ring 1.2s linear infinite;
}

@keyframes dual-ring {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
