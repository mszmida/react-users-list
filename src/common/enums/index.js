export const REQUEST_STATUS = {
  IDLE: 'IDLE',
  IN_PROGRESS: 'IN_PROGRESS',
  SUCCESS: 'SUCCESS',
  ERROR: 'ERROR'
};
