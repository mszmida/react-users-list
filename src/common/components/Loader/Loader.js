import React from 'react';
import PropTypes from 'prop-types';

import styles from './Loader.module.css';
import { REQUEST_STATUS } from '../../enums';

function Loader({ status, children }) {
  switch (status) {
    case REQUEST_STATUS.IN_PROGRESS:
      return (
        <div className={styles.component}>
          <div className={styles.loaderDualRing}></div>
        </div>
      );

    case REQUEST_STATUS.SUCCESS:
      return children;

    default:
      return null;
  }
}

Loader.propTypes = {
  status: PropTypes.string,
  children: PropTypes.node
};

export default Loader;
