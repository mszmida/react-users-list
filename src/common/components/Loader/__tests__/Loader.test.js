import React from 'react';
import { render } from '@testing-library/react';

import Loader from '../Loader';
import { REQUEST_STATUS } from '../../../enums';

describe('<Loader /> component', () => {
  test('renders loader instead of children for status "IN_PROGRESS"', () => {
    const childRef = React.createRef();

    const { container, asFragment } = render(
      <Loader status={REQUEST_STATUS.IN_PROGRESS}>
        <div ref={childRef}>child</div>
      </Loader>
    );

    expect(container.firstChild).not.toContainElement(childRef.current);
    expect(asFragment()).toMatchSnapshot();
  });

  test('renders children instead of loader for status "SUCCESS', () => {
    const childRef = React.createRef();

    const { container } = render(
      <Loader status={REQUEST_STATUS.SUCCESS}>
        <div ref={childRef}>child</div>
      </Loader>
    );

    expect(container.firstChild).toContainElement(childRef.current);
  });

  test('renders null for status "IDLE', () => {
    const { container } = render(
      <Loader status={REQUEST_STATUS.IDLE}>
        <div>child</div>
      </Loader>
    );

    expect(container.firstChild).not.toBeInTheDocument();
  });

  test('renders null for status "ERROR', () => {
    const { container } = render(
      <Loader status={REQUEST_STATUS.ERROR}>
        <div>child</div>
      </Loader>
    );

    expect(container.firstChild).not.toBeInTheDocument();
  });
});
