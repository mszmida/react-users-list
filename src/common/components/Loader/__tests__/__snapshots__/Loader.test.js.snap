// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loader /> component renders loader instead of children for status "IN_PROGRESS" 1`] = `
<DocumentFragment>
  <div
    class="component"
  >
    <div
      class="loaderDualRing"
    />
  </div>
</DocumentFragment>
`;
